blockchain/blockindex.go

```diff
@@ -574,14 +869,418 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
 // This function is safe for concurrent access.
 func (bi *blockIndex) UnsetStatusFlags(node *blockNode, flags blockStatus) {
 	bi.Lock()
-	origStatus := node.status
-	node.status &^= flags
-	if node.status != origStatus {
-		bi.modified[node] = struct{}{}
+	bi.unsetStatusFlags(node, flags)
+	bi.Unlock()
+}
+
+// addBestChainCandidate adds the passed block node as a potential candidate
+// for becoming the tip of the best chain.
+//
+// This function MUST be called with the block index lock held (for writes).
+func (bi *blockIndex) addBestChainCandidate(node *blockNode) {
+	bi.bestChainCandidates[node] = struct{}{}
+}
+
+// pruneCachedTips removes old cached chain tips used to optimize block
+// invalidation by treating the passed best known block as a reference point.
+//
+// This function MUST be called with the block index lock held (for writes).
+func (bi *blockIndex) pruneCachedTips(bestNode *blockNode) {
+	// No blocks exist before height 0.
+	height := bestNode.height - cachedTipsPruneDepth
+	if height <= 0 {
+		return
+	}
+
+	for hash, n := range bi.cachedTips {
+		if n.height < height {
+			delete(bi.cachedTips, hash)
+		}
+	}
+	bi.cachedTipsStart = height
+	bi.cachedTipsLastPruned = time.Now()
```

Comment:
I'll update it, but, I actually didn't bother because it really doesn't matter. This method is only ever called when block is connected and the height can only be <= 0 for the first 12 blocks and thus would only be hit during the initial sync and/or the first blocks of simnet anyway.

Reply:
Updated.